src/webview/html/messageAsHtml.js

```diff
@@ -72,13 +78,90 @@ $!${messageReactionListAsHtml(reactions, ownUser.user_id, allImageEmojiById)}
 `;
 };
 
-const widgetBody = (message: Message | Outbox) => template`
+/**
+ * Render the body of a message that has submessages.
+ *
+ * Must not be called on a message without any submessages.
+ */
+const widgetBody = (message: Message, ownUserId: UserId) => {
+  invariant(
+    message.submessages !== undefined && message.submessages.length > 0,
+    'should have submessages',
+  );
+
+  const sortedSubmessages = message.submessages.concat();
+  sortedSubmessages.sort((m1, m2) => m1.id - m2.id);
+
+  const widgetSubmessages: $ReadOnlyArray<{
+    sender_id: number,
+    content: SubmessageData,
+  }> = sortedSubmessages
+    .filter(submessage => submessage.msg_type === 'widget')
+    .map(submessage => ({
+      sender_id: submessage.sender_id,
+      content: JSON.parse(submessage.content),
+    }));
+
+  const pollWidget = widgetSubmessages[0];
+  if (!pollWidget || !pollWidget.content || pollWidget.content.widget_type !== 'poll') {
+    return template`
 $!${message.content}
 <div class="special-message"
  ><p>Interactive message</p
  ><p>To use, open on web or desktop</p
 ></div>
 `;
+  }
+
+  if (pollWidget.content.extra_data == null) {
+    // We don't expect this to happen in general, but there are some malformed
+    // messages lying around that will trigger this.
+    return template`$!${message.content}`;
+  }
+
+  const pollData = new PollData({
+    message_sender_id: message.sender_id,
+    current_user_id: ownUserId,
+    is_my_poll: message.sender_id === ownUserId,
+    question:
+      pollWidget.content.extra_data.question == null ? pollWidget.content.extra_data.question : '',
+    options:
+      pollWidget.content.extra_data.options == null ? pollWidget.content.extra_data.options : [],
+    // TODO: Implement this.
+    comma_separated_names: () => '',
+    report_error_function: (msg: string) => {
+      logging.error(msg);
+    },
+  });
+
+  for (const pollEvent of widgetSubmessages) {
+    pollData.handle_event(pollEvent.sender_id, pollEvent.content);
```

Comment:
> I hadn't realized that PollData was robust to receiving the initial `widget` submessage in addition to the "event" submessages, but it seems that it is.

Hmm yeah, it looks like it is. It looks for `type` and drops anything that doesn't have that, and the initial submessage doesn't have it.

I'm not sure that's really part of its interface, though -- it seems like a bit of an odd thing to do. The web caller doesn't re-pass it that first submessage; in `activate` in `widgetize.js` the flow looks like this:
```js
    events.shift();

    // …

    // Replay any events that already happened.  (This is common
    // when you narrow to a message after other users have already
    // interacted with it.)
    if (events.length > 0) {
        widget_elem.handle_events(events);
    }
```

So that `shift` drops the first element so that it doesn't go to the `handle_events` call. (Which in turn is basically a loop with `handle_event`.)

It's pretty easy to similarly skip the first element from what we use for `handle_event`, and that makes it a bit clearer to understand what's going on, so let's do that.

Reply:
Oh, wait, that's because I got the conditional backwards, oops :disappointed: 

Which means that the flow error I was fighting with before is real :/